File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridge = [0] * bridge_length
    
    bridge = deque(bridge)
    truck_weights = deque(truck_weights)
    
    weight_sum = 0
    while len(bridge) != 0:
        answer += 1
        
        bridge_front = bridge.popleft()
        if bridge_front != 0:
            weight_sum -= bridge_front
        
        if truck_weights:
            if weight_sum + truck_weights[0] <= weight:
                into_truck = truck_weights.popleft()
                weight_sum += into_truck
                bridge.append(into_truck)
            else:
                bridge.append(0)
    return answer

File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int currentWeight = 0;
        Queue<Integer> bridge = new LinkedList<>();
        Queue<Integer> truckQueue = new LinkedList<>();

        for (int truckWeight : truck_weights) {
            truckQueue.offer(truckWeight);
        }

        for (int i = 0; i < bridge_length; i++) {
            bridge.offer(0);
        }

        while (!truckQueue.isEmpty()) {
            int currentTruck = truckQueue.peek();

            if (currentWeight + currentTruck - bridge.peek() <= weight) {
                currentWeight += currentTruck - bridge.poll();
                bridge.offer(truckQueue.poll());
            } else {
                currentWeight -= bridge.poll();
                bridge.offer(0);
            }

            answer++;
        }

        answer += bridge_length;

        return answer;
    }
}
